### ISSUE TITLE
Multicast: Streams do not reconnect on error

### ISSUE THREAD

**Issue by tormoder** [relab/gorums#24]
Original issue by @s111 was relab/gorums-dev#31.

**s111 commented on Feb 27**
Streams are closed on any error and they do not reconnect. If a server ever crashes, is stopped, etc., and recovers, it won't receive multicast messages that were intended for it.

```
+--------+      +--------+
| server | <=== | server |
|    1   | ===> |    2   |
+--------+      +--------+

Server 2 crashes and recovers.

+--------+      +--------+
| server | <=== | server |
|    1   | =X   |    2   |
+--------+      +--------+

Connection from server 1 to server 2 is now permanently broken.
```

￼**meling commented on Feb 27**
Just to clarify. Do you mean between the client and one of the servers? That is, does this occur when the client does a `WriteAsync` on the `Configuration` object? (referring to the default register implementation in the dev folder). Can you provide some more details on how to reproduce this?
￼￼  ￼  ￼
**s111 commented on Feb 27**
This is how gRPC defines the `Stream` interface:

```
// Stream defines the common interface a client or server stream has to satisfy.
type Stream interface {
	// Context returns the context for this stream.
	Context() context.Context
	// SendMsg blocks until it sends m, the stream is done or the stream
	// breaks.
	// On error, it aborts the stream and returns an RPC status on client
	// side. On server side, it simply returns the error to the caller.
	// SendMsg is called by generated code. Also Users can call SendMsg
	// directly when it is really needed in their use cases.
	SendMsg(m interface{}) error
	// RecvMsg blocks until it receives a message or the stream is
	// done. On client side, it returns io.EOF when the stream is done. On
	// any other error, it aborts the stream and returns an RPC status. On
	// server side, it simply returns the error to the caller.
	RecvMsg(m interface{}) error
}
```

The important part being:
On any other error, it aborts the stream and returns an RPC status.

Basically if any error occur during a `RecvMsg` call, the underlying stream will be closed and further calls to `RecvMsg` will just return an error. That is, streams do not implement the same backoff and try again later mechanism that regular gRPC calls do.
￼￼  ￼  ￼
**s111 commented on Feb 27**
The stream is created on start up when all nodes are connecting: register.pb.go#L999.
If it ever goes down, it is not re-created.
I'm bringing this up because I assumed gRPC would try to reconnect when the stream broke down. And as documented, it does not. So I don't necessarily think we need to modify the call, but it might be good to describe this behavoir in the documentation somewhere.
￼
**meling commented on Feb 27**
Just a quick mention of the TODO in the relevant file. It would probably/maybe be helpful to also get a log message when this happens. Maybe it should be added to the top-level documentation on multicast methods.

**Comment by meling** [comment]
@Raytar is this issue already resolved with the ordered implementation? And if so, will is also work for multicast calls?

**Comment by Raytar** [comment]
Yes, but it won't work for multicast unfortunately.  The [`reconnectStream`](https://github.com/relab/gorums/blob/1392225d54e77cbd7da2e15433de965ed1072cd0/cmd/protoc-gen-gorums/dev/ordering.go#L133) function implements exponential backoff for the orderedNodeStream

**Comment by meling** [comment]
Seems as it might be possible to reuse the same node stream for multicast messages as well. Leaving this issue open for now.

**Comment by Raytar** [comment]
Yeah, we could probably reimplement multicast on top of it.

**Comment by meling** [comment]
Closed by #65
